### PR TITLE
✨  refactor providerID duplicate checks

### LINKED ISSUE #589: break the duplicateProviderIDsExist function into two separate functions
providerID duplication can happen for multiple reasons. For example:
Manual modification of cloud init data
An isolated/dead node rejoining a clustes 

Due to these and other reasons, it might be good run the detection logic regularly, not only during node provisioning.

**to do:**
break the duplicateProviderIDsExist function into two as follows

```
func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string) 
func (m *MachineManager) providerIDUsedMultipleTimes(validNodes map[string][]string, providerIDLegacy, providerIDNew string)

```
**duplicateProviderIDsExist** runs regularly, including during a provisioning of a node. 
**providerIDUsedMultipleTimes** runs only during provisioning of a node.

### PR DESCRIPTION
**Pull request by Xenwar**
refactor providerID duplicate checks

ProviderID duplicate check is broken into two. 
`currentProviderIDUsedMultipleTimes` is run during provisioning of a node. If focuses on a single providerID. To see if it is being used in different formats. 

`duplicateProviderIDsExist` can be run at any time. The purpose of it is to detect duplicates introduced either manually or due to Manual modification of cloud init data an isolated/dead node rejoining a cluster.

Fixes #589

### CODE REVIEW

File: baremetal/metal3machine_manager.go [hunk 1]
```diff
@@ -1853,16 +1858,31 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
...
 }
 
-// duplicateProviderIDsExist determnes if a providerID is already in use by other nodes.
-func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string, providerIDLegacy, providerIDNew string) error {
+// duplicateProviderIDsExist determnes if a providerID is used in multiple nodes.
+// this can be called at any time.
+func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string) error {
+	for providerID, nodes := range validNodes {
+		matchingNodes := strings.Join(nodes, ",")
+		// duplicates due to the same, at least, two instances of legacy OR new OR unknown formats being used in multiple nodes.
```
Comment (smoshiur1237) [inline baremetal/metal3machine_manager.go:1866]:
```suggestion
		// verify if same providerID is used by multiple nodes.
```

**Review by smoshiur1237 — CHANGES_REQUESTED** [review]
Few changes needed for nits.  I think it should include the unit test with the same PR.

File: baremetal/metal3machine_manager.go [hunk 2]
```diff
@@ -1873,17 +1893,11 @@ func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]strin
...
-			m.Log.Info(errMsg)
-			return errors.New(errMsg)
-		}
 	}
-	// duplicates due to both the legacy AND new providerIDs being used by multiple nodes.
+	// duplicates due to both the legacy AND new providerIDs being used by different nodes.
 	if duplicateUsageCounter > 1 {
 		duplicateNodesNames := strings.Join(duplicateNodes, ",")
-		errMsg := fmt.Sprintf("both providerIDs (%s and %s) cannot be used at the same time by node: (%s)", providerIDLegacy, providerIDNew, duplicateNodesNames)
+		errMsg := fmt.Sprintf("both formats of providerID (%s and %s) cannot be used at the same time by node(s): (%s)", providerIDLegacy, providerIDNew, duplicateNodesNames)
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1900]: done

File: baremetal/metal3machine_manager.go [hunk 3]
```diff
@@ -1853,16 +1858,31 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
 	return matchingNodesCount, nil
 }
 
-// duplicateProviderIDsExist determnes if a providerID is already in use by other nodes.
-func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string, providerIDLegacy, providerIDNew string) error {
+// duplicateProviderIDsExist determnes if a providerID is used in multiple nodes.
+// this can be called at any time.
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1862]: removed

File: baremetal/metal3machine_manager.go [hunk 4]
```diff
@@ -1853,16 +1858,31 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
...
+		if len(nodes) > 1 {
+			errMsg := fmt.Sprintf("providerID %s is in use by multiple nodes: (%s)", providerID, matchingNodes)
+			m.Log.Info(errMsg)
+			return errors.New(errMsg)
+		}
+	}
+	return nil
+}
+
+// providerIDUsedMultipleTimes determnes if a providerID is already in use in different formats.
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1876]: fixed

File: baremetal/metal3machine_manager.go [hunk 5]
```diff
@@ -1853,16 +1858,31 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
...
-// duplicateProviderIDsExist determnes if a providerID is already in use by other nodes.
-func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string, providerIDLegacy, providerIDNew string) error {
+// duplicateProviderIDsExist determnes if a providerID is used in multiple nodes.
+// this can be called at any time.
+func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string) error {
+	for providerID, nodes := range validNodes {
+		matchingNodes := strings.Join(nodes, ",")
+		// duplicates due to the same, at least, two instances of legacy OR new OR unknown formats being used in multiple nodes.
+		if len(nodes) > 1 {
+			errMsg := fmt.Sprintf("providerID %s is in use by multiple nodes: (%s)", providerID, matchingNodes)
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1868]: fixed

File: baremetal/metal3machine_manager.go [hunk 6]
```diff
@@ -1853,16 +1858,31 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
 	return matchingNodesCount, nil
 }
 
-// duplicateProviderIDsExist determnes if a providerID is already in use by other nodes.
-func (m *MachineManager) duplicateProviderIDsExist(validNodes map[string][]string, providerIDLegacy, providerIDNew string) error {
+// duplicateProviderIDsExist determnes if a providerID is used in multiple nodes.
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1861]: done

File: baremetal/metal3machine_manager.go [hunk 7]
```diff
@@ -1841,7 +1841,12 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
 			validNodes[providerIDOnNode] = append(validNodes[providerIDOnNode], node.GetName())
 		}
 	}
-	err = m.duplicateProviderIDsExist(validNodes, providerIDLegacy, providerIDNew)
+	err = m.duplicateProviderIDsExist(validNodes)
+	if err != nil {
+		// There are, at least, two nodes. Details are in err
```
[the remark this answers is not in the document]
Comment (Xenwar) [reply baremetal/metal3machine_manager.go:1846]: done

**Comment by Xenwar** [comment]
@maelk If you have a comment on this PR. 
My thinking when creating this PR is to avoid provider-ID duplication that is introduced any way after the provisioning of a node.  Currently, we are preventing duplicate usage during provisioning of a node.

This PR breaks the check into two. 
1. duding provisioning 
2. regularly at intervals 

I would like to know if there are use cases that fit my assumption that:
- an isolated node using `metal3://bmh-id1` joins back the cluster after a while and the `metal3://bmh-id1` is used on a new node. 

or if there is any other use case that requires such duplicate detection. 
Thanks.

**Comment by Xenwar** [comment]
PR no longer needed as other PRs have handled parts of it.

**Pull request closed without merging**